src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker_Patterns.cs

```diff
@@ -246,13 +247,66 @@ protected override void VisitSwitchSection(BoundSwitchSection node, bool isLastS
             VisitStatementList(node);
         }
 
+        private struct PossiblyConditionalState
+        {
+            public LocalState State;
+            public LocalState StateWhenTrue;
+            public LocalState StateWhenFalse;
+            public bool IsConditionalState;
+
+            public PossiblyConditionalState(LocalState state)
+            {
+                StateWhenTrue = StateWhenFalse = default;
+                IsConditionalState = false;
+                State = state.Clone();
+            }
+
+            public PossiblyConditionalState(NullableWalker nullableWalker)
+            {
+                if (nullableWalker.IsConditionalState)
+                {
+                    StateWhenTrue = nullableWalker.StateWhenTrue.Clone();
+                    StateWhenFalse = nullableWalker.StateWhenFalse.Clone();
+                    IsConditionalState = true;
+                    State = default;
+                }
+                else
+                {
+                    StateWhenTrue = StateWhenFalse = default;
+                    IsConditionalState = false;
+                    State = nullableWalker.State.Clone();
+                }
```

Comment:
These cases are duplicated in the constructor above and `Clone()` below. Consider adding another constructor, `PossiblyConditionalState(LocalState whenTrue, LocalState whenFalse)`, and change this method to a `static` factory method that calls one of the two constructors depending on `IsConditionalState`.